File: lib/src/vault/seed_vault.dart
// O cofre: onde a seed cifrada mora no aparelho.
//
// Contrato do projeto — vale para sempre:
//   - só entra aqui o blob CIFRADO (ver pin_crypto.dart), nunca a frase em claro;
//   - nada daqui sai para rede, log, analytics ou relatório de erro;
//   - a interface existe para que o teste rode sem aparelho e para que uma
//     segunda implementação (ex.: outra rede) reaproveite o mesmo fluxo.

import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import 'pin_crypto.dart';

/// Tentativas de PIN antes do cofre se apagar.
///
/// Apagar NÃO é perder o dinheiro: a carteira volta pelas 12 palavras. É o que
/// impede alguém com o aparelho na mão de ficar tentando PIN à vontade.
const int kMaxPinAttempts = 10;

abstract class SeedVault {
  Future<bool> hasSeed();

  /// Grava a frase cifrada com o PIN. Sobrescreve o que houver.
  Future<void> saveSeed(String mnemonic, String pin);

  /// Devolve a frase. Lança [WrongPinException] quando o PIN não abre.
  Future<String> unlock(String pin);

  /// Apaga tudo. Irreversível no aparelho — só as 12 palavras trazem de volta.
  Future<void> wipe();

  /// Quantas tentativas de PIN ainda restam antes do apagamento.
  Future<int> remainingAttempts();
}

class SecureStorageSeedVault implements SeedVault {
  SecureStorageSeedVault({FlutterSecureStorage? storage})
      : _storage = storage ??
            const FlutterSecureStorage(
              // Na v11 o padrao do Android JA e Keystore com AES-GCM e chave
              // envelopada em RSA-OAEP; nao existe mais opcao a ligar aqui.
              aOptions: AndroidOptions(),
              // No iOS, a seed so e legivel neste aparelho e apos o primeiro
              // desbloqueio: nao vai para backup nem para outro dispositivo.
              iOptions: IOSOptions(
                accessibility: KeychainAccessibility.first_unlock_this_device,
              ),
            );

  final FlutterSecureStorage _storage;

  static const String _seedKey = 'bitibridge.seed.v1';
  static const String _failsKey = 'bitibridge.seed.fails.v1';

  @override
  Future<bool> hasSeed() async => await _storage.read(key: _seedKey) != null;

  @override
  Future<void> saveSeed(String mnemonic, String pin) async {
    final EncryptedSeed blob = await encryptSeedWithPin(mnemonic, pin);
    await _storage.write(key: _seedKey, value: blob.encode());
    await _storage.write(key: _failsKey, value: '0');
  }

  @override
  Future<String> unlock(String pin) async {
    final String? raw = await _storage.read(key: _seedKey);
    if (raw == null) {
      throw StateError('Nenhuma carteira guardada neste aparelho');
    }
    try {
      final String mnemonic =
          await decryptSeedWithPin(EncryptedSeed.decode(raw), pin);
      await _storage.write(key: _failsKey, value: '0');
      return mnemonic;
    } on WrongPinException {
      final int fails = await _readFails() + 1;
      if (fails >= kMaxPinAttempts) {
        await wipe();
      } else {
        await _storage.write(key: _failsKey, value: '$fails');
      }
      rethrow;
    }
  }

  @override
  Future<void> wipe() async {
    await _storage.delete(key: _seedKey);
    await _storage.delete(key: _failsKey);
  }

  @override
  Future<int> remainingAttempts() async =>
      kMaxPinAttempts - await _readFails();

  Future<int> _readFails() async =>
      int.tryParse(await _storage.read(key: _failsKey) ?? '0') ?? 0;
}

File: src/lib/vault/seedVault.ts
// O cofre: onde a seed cifrada mora no navegador.
//
// Contrato do projeto — vale para sempre:
//   - só entra aqui o blob CIFRADO (ver pinCrypto.ts), nunca a frase em claro;
//   - nada daqui sai para rede, log ou relatório de erro;
//   - a interface existe para que o teste rode sem navegador e para que o
//     armazenamento possa mudar sem mexer nas telas.

import {
  decodeEncryptedSeed,
  decryptSeedWithPin,
  encodeEncryptedSeed,
  encryptSeedWithPin,
  WrongPinError,
} from "./pinCrypto";

/**
 * Tentativas de PIN antes de o cofre se apagar.
 *
 * Apagar NÃO é perder o dinheiro: a carteira volta pelas 12 palavras. É o que
 * impede alguém com o navegador aberto de ficar tentando PIN à vontade.
 */
export const MAX_PIN_ATTEMPTS = 10;

export interface SeedVault {
  hasSeed(): Promise<boolean>;
  saveSeed(mnemonic: string, pin: string): Promise<void>;
  /** Devolve a frase. Lança WrongPinError quando o PIN não abre. */
  unlock(pin: string): Promise<string>;
  /** Apaga tudo. Só as 12 palavras trazem de volta. */
  wipe(): Promise<void>;
  remainingAttempts(): Promise<number>;
}

/** Armazenamento mínimo de chave/valor — o localStorage do navegador o satisfaz. */
export interface KeyValueStore {
  getItem(key: string): string | null;
  setItem(key: string, value: string): void;
  removeItem(key: string): void;
}

const SEED_KEY = "bitibridge.seed.v1";
const FAILS_KEY = "bitibridge.seed.fails.v1";

export class BrowserSeedVault implements SeedVault {
  constructor(private readonly store: KeyValueStore = localStorage) {}

  async hasSeed(): Promise<boolean> {
    return this.store.getItem(SEED_KEY) !== null;
  }

  async saveSeed(mnemonic: string, pin: string): Promise<void> {
    const blob = await encryptSeedWithPin(mnemonic, pin);
    this.store.setItem(SEED_KEY, encodeEncryptedSeed(blob));
    this.store.setItem(FAILS_KEY, "0");
  }

  async unlock(pin: string): Promise<string> {
    const raw = this.store.getItem(SEED_KEY);
    if (raw === null) throw new Error("Nenhuma carteira guardada neste navegador");
    try {
      const mnemonic = await decryptSeedWithPin(decodeEncryptedSeed(raw), pin);
      this.store.setItem(FAILS_KEY, "0");
      return mnemonic;
    } catch (error) {
      if (!(error instanceof WrongPinError)) throw error;
      const fails = this.readFails() + 1;
      if (fails >= MAX_PIN_ATTEMPTS) {
        await this.wipe();
      } else {
        this.store.setItem(FAILS_KEY, String(fails));
      }
      throw error;
    }
  }

  async wipe(): Promise<void> {
    this.store.removeItem(SEED_KEY);
    this.store.removeItem(FAILS_KEY);
  }

  async remainingAttempts(): Promise<number> {
    return MAX_PIN_ATTEMPTS - this.readFails();
  }

  private readFails(): number {
    const raw = this.store.getItem(FAILS_KEY);
    const parsed = raw === null ? 0 : Number.parseInt(raw, 10);
    return Number.isFinite(parsed) ? parsed : 0;
  }
}
